ktor-client/ktor-client-cio/jvm/src/io/ktor/client/engine/cio/ConnectionPipeline.kt

```diff
@@ -69,50 +69,55 @@ internal class ConnectionPipeline(
-                            val proxyChannel = ByteChannel()
-                            skipTask = skipCancels(responseChannel, proxyChannel)
-                            proxyChannel
-                        } else ByteReadChannel.Empty
-
-                        val response = CIOHttpResponse(
-                            task.request, headers, requestTime,
-                            body,
-                            rawResponse,
-                            coroutineContext = callContext
-                        )
-
-                        task.response.complete(response)
-
-                        responseChannel?.use {
-                            parseHttpBody(
-                                contentLength,
-                                transferEncoding,
-                                connectionType,
-                                networkInput,
-                                this
-                            )
-                        }
-                        skipTask?.join()
-                    } finally {
+                    val callContext = task.context
+                    val callJob = callContext[Job]!!
+                    callJob.invokeOnCompletion {
                         rawResponse.headers.release()
                     }
+
+                    val method = task.request.method
+                    val contentLength = rawResponse.headers[HttpHeaders.ContentLength]?.toString()?.toLong() ?: -1L
+                    val transferEncoding = rawResponse.headers[HttpHeaders.TransferEncoding]
+                    val chunked = transferEncoding == "chunked"
+                    val connectionType = ConnectionOptions.parse(rawResponse.headers[HttpHeaders.Connection])
+                    val headers = CIOHeaders(rawResponse.headers)
+
+                    shouldClose = (connectionType == ConnectionOptions.Close)
+
+                    val hasBody = (contentLength > 0 || chunked) && method != HttpMethod.Head
+                    val responseChannel = if (hasBody) ByteChannel() else null
+
+                    var skipTask: Job? = null
+                    val body: ByteReadChannel = if (responseChannel != null) {
+                        val proxyChannel = ByteChannel()
+                        skipTask = skipCancels(responseChannel, proxyChannel)
+                        proxyChannel
+                    } else ByteReadChannel.Empty
+
+                    callJob.invokeOnCompletion {
+                        body.cancel()
+                    }
+
+                    val response = CIOHttpResponse(
+                        task.request, headers, requestTime,
+                        body,
+                        rawResponse,
+                        coroutineContext = callContext
+                    )
+
+                    task.response.complete(response)
+
+                    responseChannel?.use {
+                        parseHttpBody(
```

Comment:
what if fails with an exception here? would/should it lead to all calls failure?